### ISSUE TITLE
Add a skill optimizer (/skill-optimizer) that tunes SKILL.md bodies against benchmarks

### ISSUE THREAD

**Issue by bostonaholic** [bostonaholic/team#39]
## Summary

Add a **skill optimizer** — `/skill-optimizer <skill-name>` — that treats a
SKILL.md **body** as the trainable parameters of an otherwise-frozen agent and
tunes it against a benchmark. The optimizer runs the skill on realistic tasks,
proposes specific edits from what worked and what didn't, re-tests, and **keeps
only changes that measurably improve the score**. The agent never changes; only
the skill document does.

It's gradient descent by analogy: the SKILL.md body is the weights, the
benchmark is the loss, an LLM reflection step is the "gradient" (concrete
edits), a learning-rate budget clips how many edits land per step, and a hard
validation gate is the optimizer that accepts or rejects each step.

## Motivation

- Skill bodies in this plugin are never empirically tuned. We
  have `create-team-skill` to *write* a skill, but nothing to *improve* one
  against measured outcomes.
- We already have the substrate: the behavioral evals harness (`tests/` +
  `evals/`) spawns `claude -p`, parses NDJSON transcripts, runs deterministic
  judges, and tracks variance/regression in `eval-store.ts`. A skill optimizer
  is a natural new consumer of exactly that machinery.
- "Did this rewrite actually help?" is a vibe today. A closed loop makes it a
  number — with a gate strict enough that LLM-judge noise can't sneak through.

## Scope of mutation (the core safety property)

**Body-only.** The optimizer edits the methodology/prose body of a SKILL.md and
**never** touches frontmatter (`description:`, etc.). Triggering is a separate
concern with separate failure modes; freezing the routing contract the rest of
the pipeline depends on is what makes every optimization step safe to accept.
Edits that cross the frontmatter boundary or land inside a code fence are
rejected, not applied.

## Edit model

Edits are **anchored patch ops**, not freeform rewrites (a `--rewrite` mode is
an option, but `--patch` is the default):

- `add` — insert `content` at an `anchor`
- `replace` — swap `target` for `replacement`
- `delete` — remove `target`

Applying an edit returns a **tagged result** (`applied` | `rejected`) rather
than throwing. Rejections carry a reason (`anchor_not_found`,
`anchor_ambiguous`, `target_ambiguous`, `inside_code_fence`,
`crosses_frontmatter`, `no_change`, …) and go into a bounded rejected-edits
buffer so the optimizer doesn't re-propose dead ends across steps or resumes.

## Benchmark + judges

A JSONL file, **one task per line**, each `{task_id, task, judge}`. Lives at
`evals/fixtures/skills/<name>/skill-optimizer-benchmark.jsonl` to match our fixtures
convention. Loading is truncation-tolerant: finished lines are kept, only a
truncated trailing line is dropped.

Judge kinds:

- **`rule`** (deterministic, free) — a list of checks: `contains`, `regex`,
  `section_present`, `max_chars`, `min_citations`, `tool_called`,
  `tool_not_called`. This is the core metric — no LLM judge needed.
- **`llm`** — a rubric scored by a model, for quality dimensions rules can't capture.
- **`qrels`** — expected output set + `k`, for routing/dispatch-style skills.

Example line:

```json
{"task_id":"t-001","task":"...","judge":{"kind":"rule","checks":[{"op":"max_chars","arg":1500},{"op":"contains","arg":"worktree"}]}}
```

## The optimization loop

**Split** the benchmark into **train : select : test** (default `4:1:5`; `1:1:1`
for tiny starters). A floor of **D_sel ≥ 5** is enforced at load — meaningful
validation needs enough select tasks, so the loader refuses below it. Train
drives proposals, *select* gates acceptance, *test* is reported but never
optimized against.

Then, for each step (`--epochs` outer, default 4; `--batch-size` inner, default 8):

1. **Forward pass.** Run the candidate skill against a train batch, capturing
   trajectories (final text + tool calls + usage) in-process — never written to
   any persistent store.
2. **Backward pass.** **Two reflect calls** — one over the failures, one over
   the successes — propose edits addressing what didn't work and reinforcing
   what did.
3. **Rank + clip.** Take the top-N edits within the learning-rate budget. `--lr`
   (default 4) caps edits per step; `--lr-schedule` (default `cosine`, also
   `linear`/`constant`) decays the budget over epochs so later steps make
   smaller, safer changes.
4. **Apply.** Patch the body via tagged edit ops; rejects go to the buffer.
5. **Validation gate (non-negotiable).** Every *select* task runs the judge
   **3 times** and takes the **median**. A candidate is accepted **only if its
   median beats the prior best by more than ε = 0.05** — otherwise rejected.
6. **Commit.** History-intent-first atomic write so a crash never leaves a
   half-written SKILL.md.

## Three model roles

| Role | Default tier | Job |
|---|---|---|
| optimizer-model | deep | reflect + propose edits |
| target-model | subagent | execute the skill on tasks |
| judge-model | reasoning | score rollouts |

Each overridable (`--optimizer-model`, `--target-model`, `--judge-model`).

## Bootstrapping a benchmark

Writing a benchmark by hand is the adoption blocker, so generate a starter:

```
/skill-optimizer <skill-name> --bootstrap-from-skill
```

- One LLM call reads `skills/<name>/SKILL.md`, infers what the skill is supposed
  to produce, and writes ~15 tasks (tunable `--bootstrap-tasks N`, max 50) with
  deterministic rule judges to `skill-optimizer-benchmark.jsonl`.
- Generated judges are **weak drafts** (generic `contains`, loose length caps).
  The file is written with a `# BOOTSTRAP_PENDING_REVIEW` sentinel, and the
  optimizer **refuses to run until that line is deleted** (re-run with
  `--bootstrap-reviewed`) — a human must strengthen the judges first. This also
  blocks a skill from gaming a benchmark it wrote about itself.
- Any task left with fewer than two valid checks is dropped whole.

Hand-authored benchmarks remain the fallback for skills the generator can't
handle well.

## Artifacts

```
evals/fixtures/skills/<name>/
  skill-optimizer-benchmark.jsonl     ← what success looks like
  skill-optimizer/
    best.md                    ← current best body
    versions/v0001_e1_s1.md …  ← per-step snapshots
    history.json               ← append-only run record
    rejected.json              ← bounded LRU of rejected edits
```

## Safety guards

| Guard | Prevents |
|---|---|
| Mandatory median-of-3 + ε=0.05 validation gate | accepting judge noise as improvement |
| Frontmatter mutation forbidden | routing-surface drift |
| Per-skill lock | concurrent runs corrupting history/versions |
| Bootstrap-review sentinel | self-referential benchmark gaming |
| Read-only tool sandbox during rollouts | optimization runs writing junk side effects |
| Atomic, intent-first commit | half-written SKILL.md on crash |
| Cost preflight + `--max-cost-usd` / `--max-runtime-min` caps | surprise mid-run budget blowups |
| Dirty-tree refusal (`--force` to override) | overwriting uncommitted changes |
| Bundled-skill gate (`--allow-mutate-bundled`) | silently mutating plugin-shipped skills |

## Flags (proposed)

`--benchmark`, `--bootstrap-from-skill`, `--bootstrap-tasks N`,
`--bootstrap-reviewed`, `--epochs`, `--batch-size`, `--lr`, `--lr-schedule`,
`--split`, `--optimizer-model`, `--target-model`, `--judge-model`,
`--patch|--rewrite`, `--dry-run` (cost preview, no LLM calls), `--no-mutate`
(write `proposed.md`, don't replace the skill), `--max-cost-usd`,
`--max-runtime-min`, `--force`, `--resume <run-id>`, `--json`.

Exit codes: `0` improved + accepted (or `--no-mutate` wrote `proposed.md`),
`1` no improvement (skill unchanged), `2` aborted by a gate (dirty tree, over
budget, benchmark validation, etc.).

## Fit with the existing harness

- Reuse `tests/helpers/session-runner.ts` to run the skill and parse outcomes.
- `rule` judges align with the existing `ground-truth.json` / outcome-judge pattern.
- Two-tier split: a free `*.test.ts` static gate validates benchmark JSONL +
  the pending-review sentinel + the D_sel floor; a paid `*.evals.ts` tier runs
  the optimization loop behind `EVALS_ANTHROPIC_API_KEY`.
- Reuse variance/regression tracking from `eval-store.ts`.

## Scope

**In scope (v1):** single-skill body optimization with anchored patch edits;
`rule`/`llm`/`qrels` judges; `--bootstrap-from-skill` with the review sentinel;
train/select/test split + D_sel floor; epochs/batch/lr + cosine schedule;
median-of-3 + ε gate; rejected-edit buffer; versions/best/history layout;
per-skill lock; cost preflight; reviewable diff before write-back.

**Out of scope (v1) — follow-ups:** cross-skill batch mode (`--all`) with
plugin-wide cost caps; cross-model fleet optimization (`--target-models`);
background/queued execution; mocked-write capture for write-flavored skills;
slow meta-edit on a no-improvement epoch; frontmatter/triggering optimization
(deliberately excluded — different problem).

## When NOT to use it

- No real benchmark — optimizing against guesses is worse than not optimizing.
- Tiny benchmarks (<10 tasks; D_sel<5 refuses).
- Skills whose whole job is side-effecting writes (no v1 read-only sandbox path).

## Acceptance criteria

- [ ] `--bootstrap-from-skill` writes JSONL with a `# BOOTSTRAP_PENDING_REVIEW` sentinel.
- [ ] Optimizer refuses to run while the sentinel is present; `--bootstrap-reviewed` proceeds.
- [ ] Tasks with fewer than two valid checks are dropped at load; D_sel<5 is refused.
- [ ] Edits are anchored ops; frontmatter-crossing / in-code-fence / ambiguous edits are rejected to the buffer, not applied.
- [ ] Each step runs forward (train batch) → two reflect calls → top-N within lr budget → apply.
- [ ] Validation gate: median-of-3 on select, accept only on >0.05 improvement over prior best.
- [ ] Rejected edits persist across steps and resumes.
- [ ] Concurrent runs on the same skill serialize via a per-skill lock.
- [ ] Atomic commit leaves no half-written SKILL.md on crash; `--resume` continues a run.
- [ ] Cost preflight + `--max-cost-usd` cap; `--dry-run` previews cost with no LLM calls.
- [ ] Output: test-split score, accepted-edit history, and a reviewable SKILL.md diff.
